src/Symfony/Bundle/FrameworkBundle/Form/LoginType.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Form;
+
+use Symfony\Bundle\SecurityBundle\Security\FirewallMap;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\Extension\Core\Type\PasswordType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormError;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationUtils;
+use Symfony\Component\Translation\TranslatorInterface;
+
+class LoginType extends AbstractType
+{
+    private $requestStack;
+    private $firewallMap;
+    private $authenticationUtils;
+    private $urlGenerator;
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    public function __construct(RequestStack $requestStack, FirewallMap $firewallMap, AuthenticationUtils $authenticationUtils, UrlGeneratorInterface $urlGenerator, TranslatorInterface $translator)
+    {
+        $this->requestStack = $requestStack;
+        $this->firewallMap = $firewallMap;
+        $this->authenticationUtils = $authenticationUtils;
+        $this->urlGenerator = $urlGenerator;
+        $this->translator = $translator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $firewallConfig = $this->firewallMap->getFirewallConfig($this->requestStack->getCurrentRequest());
+        if (!$firewallConfig || !$firewallConfig->isSecurityEnabled()) {
+            throw new \LogicException('You cannot use LoginType if security is not enabled');
+        }
+
+        $firewallOptions = $firewallConfig->getOptions()['form_login'];
```

Comment:
Is it safe read this key? IMHO we can get undefined index when `form_login` not defined in current firewall.

Reply:
Yes, it's unreliable, see my other comment. I think this kind of check belongs in the extension (or a compiler pass).

Reply:
I also think it's not the place the LoginType should look upon. Instead, adding a `firewall` option in the OptionsResolver would allow perfect standalone checks over the FirewallMap at runtime, check if the called firewall has enabled security, retrieve its user provider, check that the current request matches the same firewall (to avoid login be used on a different firewall that may corrupt authentication) etc.